Custom schematic returning an Observable fails with 'tree.branch is not a function' error

<!--
IF YOU DON'T FILL OUT THE FOLLOWING INFORMATION YOUR ISSUE MIGHT BE CLOSED WITHOUT INVESTIGATING
-->
### Bug Report or Feature Request (mark with an `x`)
```
- [ X] bug report -> please search issues before submitting
- [ ] feature request
```

### Area
```
- [ ] devkit
- [X ] schematics
```
### Versions
<!--
Output from: `node --version` and `npm --version`.
  Windows (7/8/10). Linux (incl. distribution). macOS (El Capitan? Sierra? High Sierra?)
-->
v6.10.1/3.10.10/Windows 10

### Repro steps
<!--
Simple steps to reproduce this bug.
Please include: commands run, packages added, related code changes.
A link to a sample repo would help too.
-->
1. Create custom schematic project, e.g. schematics @schematics/schematics:schematic --name my-schematic-collection
2. Change to my-schematic-collection folder
3. npm install
4. npm install --save rxjs
5. modify example code for my-schematic to return an Observable<Tree> instead of Tree object
    e.g. import "Observable" from "rxjs/Observable"; return Observable.of(tree);
6. npm run build
7. npm link (just to use it locally for now)
8. run schematics via following command: schematics my-schematic-collection:my-schematic
    This results in error: tree.branch is not a function

### The log given by the failure
<!-- Normally this include a stack trace and some more information. -->


### Desired functionality
<!--
What would like to see implemented?
What is the usecase?
-->
This should work (hopefully).

### Mention any other details that might be useful
<!-- Please include a link to the repo if this is related to an OSS project. -->

WORK-AROUND found:
in global installation of the devkit, change the following in:
npm\node_modules\@angular-devkit\schematics\src\engine\schematic.js

from: if (result instanceof Observable_1.Observable) {
to: if (Symbol.observable in result) {

this issue most likely is related to the fact that instanceof does not work for objects passed between different javascript contexts.  This project already seems to be using Symbol.observable check in other places.

I tried to delete the `node_modules/@angular-devkit` folder (which contains three packages: schematics, core, build-optimizer), copy over from system global `@angular-devkit` built into my project `node_modules` folder. It gets rid of the error `tree.branch is not a function` but introduce a new one. 

```
module.js:538
    throw err;
    ^

Error: Cannot find module '@angular-devkit/schematics'
    at Function.Module._resolveFilename (module.js:536:15)
    at Function.Module._load (module.js:466:25)
    at Module.require (module.js:579:17)
    at require (internal/module.js:11:18)
    at Object.<anonymous> (_PATH_TO_MY_PROJECT_/node_modules/@angular/cli/utilities/schematics.js:9:22)
    at Module._compile (module.js:635:30)
    at Object.Module._extensions..js (module.js:646:10)
    at Module.load (module.js:554:32)
    at tryModuleLoad (module.js:497:12)
    at Function.Module._load (module.js:489:3)
```

there is a bug somewhere I guess. Thanks for looking into it. @hansl

--- a/packages/angular_devkit/schematics/src/engine/schematic.ts
+++ b/packages/angular_devkit/schematics/src/engine/schematic.ts
@@ -9,6 +9,7 @@ import { BaseException } from '@angular-devkit/core';
 import { Observable } from 'rxjs/Observable';
 import 'rxjs/add/observable/of';
 import 'rxjs/add/operator/concatMap';
+import { callRule } from '../rules/call';
 import { Tree } from '../tree/interface';
 import {
   Collection,
@@ -48,18 +49,8 @@ export class SchematicImpl<CollectionT extends object, SchematicT extends object
     parentContext?: Partial<TypedSchematicContext<CollectionT, SchematicT>>,
   ): Observable<Tree> {
     const context = this._engine.createContext(this, parentContext);
+    const transformedOptions = this._engine.transformOptions(this, options);
 
-    return host.concatMap(tree => {
-      const transformedOptions = this._engine.transformOptions(this, options);
-
-      const result = this._factory(transformedOptions)(tree, context);
-      if (result instanceof Observable) {
-        return result;
-      } else if (result === undefined) {
-        return Observable.of(tree);
-      } else {
-        return Observable.of(result);
-      }
-    });
+    return callRule(this._factory(transformedOptions), host, context);
   }
 }

--- a/packages/angular_devkit/schematics/src/rules/call.ts
+++ b/packages/angular_devkit/schematics/src/rules/call.ts
@@ -81,8 +81,6 @@ export function callRule(rule: Rule,
       return Observable.of(result as Tree);
     } else if (Symbol.observable in result) {
       return result as Observable<Tree>;
-    } else if (result === undefined) {
-      return Observable.of(inputTree);
     } else {
       throw new InvalidRuleResultException(result);
     }

--- a/packages/angular_devkit/schematics/testing/schematic-test-runner.ts
+++ b/packages/angular_devkit/schematics/testing/schematic-test-runner.ts
@@ -20,9 +20,6 @@ import {
 import { Observable } from 'rxjs/Observable';
 
 
-export interface SchematicSchemaT {}
-
-
 export class UnitTestTree extends DelegateTree {
   get files() {
     const result: string[] = [];
@@ -44,10 +41,7 @@ export class SchematicTestRunner {
     this._logger = new logging.Logger('test');
     this._registry = new schema.JsonSchemaRegistry();
 
-    this._engineHost.registerOptionsTransform((
-      schematicDescription: {},
-      opts: SchematicSchemaT,
-    ) => {
+    this._engineHost.registerOptionsTransform((schematicDescription: {}, opts: object) => {
       const schematic: FileSystemSchematicDesc = schematicDescription as FileSystemSchematicDesc;
 
       if (schematic.schema && schematic.schemaJson) {
@@ -68,7 +62,7 @@ export class SchematicTestRunner {
 
   get logger(): logging.Logger { return this._logger; }
 
-  runSchematicAsync(
+  runSchematicAsync<SchematicSchemaT>(
     schematicName: string,
     opts?: SchematicSchemaT,
     tree?: Tree,
@@ -80,7 +74,11 @@ export class SchematicTestRunner {
       .map(tree => new UnitTestTree(tree));
   }
 
-  runSchematic(schematicName: string, opts?: SchematicSchemaT, tree?: Tree): UnitTestTree {
+  runSchematic<SchematicSchemaT>(
+    schematicName: string,
+    opts?: SchematicSchemaT,
+    tree?: Tree,
+  ): UnitTestTree {
     const schematic = this._collection.createSchematic(schematicName);
 
     let result: UnitTestTree | null = null;

